### DISCUSSION TITLE
Check if media file exists in Public to prevent 404 Not Found error

### DISCUSSION THREAD

**Discussion by markbruk** [quasarframework/quasar#10578] [General - Components / Directives / etc] [answered]
I am playing video files which are stored in my public folder and selected using a computed property.

The app is to be compiled for SPA, mobile (Cordova) and Electron.

I cannot seem to determine if a file exists so that I can provide an error response to the user and let them proceed to the next lesson.

Any help would be greatly appreciated.

HTML
<video 
            :key="videoSource"
            v-on:ended="vidEnded"
            v-on:play="vidRate"
            duration
            ref="vid"
            width="100%"
            height="100%"
            playsinline="true"
            webkit-playsinline="true"
            autoplay
            :poster="videoPoster"
          >
            <source 
              :src="computedVideos"
              type="video/mp4"
            />
</video>

JAVASCRIPT
computedVideos() {
      
      if (typeof this.playingNow === 'undefined') this.playingNow = 'phoneme'

      // check for undefined array element
      if (Object.keys(this.phonemes).length == 0) {
        console.log('COMPUTED VIDEOS - undefined')
      } else {
        let answer = null
        if (this.playingNow == 'phoneme') {
          answer = this.phonemes[this.selectedPhoneme].alpha
          if (answer == '') {
            this.playingNow = 'word'
            answer = this.words[this.selectedWord].word  
          }
        } else { 
          answer = this.words[this.selectedWord].word
        }
        console.log('COMPUTED VIDEOS - ANSWER variable is ', answer)

        if (this.sideView) {
          console.log('sideView')
          this.videoSource = `words_m4v/` + this.playingNow + `_` + answer + `_side.m4v#t=0.1`
          this.videoHeadIcon = `images/face_front_view.png`
          this.videoPoster = `images/simone_side_view.png`
        } else {
          console.log('frontView')
          this.videoSource = `words_m4v/` + this.playingNow + `_` + answer + `_front.m4v#t=0.1`
          this.videoHeadIcon = `images/face_side_view.png`
          this.videoPoster = `images/simone_front_view.png`
        }
      }

      // check whether to show chapter or video
      console.log('lastModule')
      let y = this.$q.localStorage.getItem('lastModule')
      if (y == '' || y == null) {
        this.$q.localStorage.set('lastModule', 'video')
        this.showVideo()
      } else {
        if (y == 'video' || y == 'dictionary' || y == 'vpa') this.showVideo()
        if (y == 'chapter') this.showChapter()
      }

      return this.videoSource
},

**Comment by pikil** [comment, marked as answer]
@markbruk If you use axios, you can utilise it's ```catch``` method:

```
api.get(
    'non-existent-video'
  ).then((response) => {
    // video exists
  }).catch((error) => {
    // Video does not exist
    console.log(error.response)
    
    if (error.response.status === 404)
      router.push('...')
      // Redirect somewhere or show another lesson, etc.
  })
```
If you do not use axios, you can get the similar result with fetch:
https://developer.mozilla.org/en-US/docs/Web/API/fetch#examples

* In case you would rather like to check the existence of files on build stage instead, see the answers below.

**Reply by markbruk** [reply]
Thank you pikil for taking the time to answer and to write your answer out.  The axios call does let me react to whether the video exists or not.

**Comment by ontwikkelfabriek** [comment]
@markbruk 

> I cannot seem to determine if a file exists so that I can provide an error response to the user and let them proceed to the next lesson.

Don't you know at `build` time what video's exist in the public folder? Sounds like you could use a (webpack) script at build time to list the video files in the public folder, and for example create a .json file with all the video's available.

**Reply by markbruk** [reply]
Yes I do know at build time what is in the folder, but I was unaware of webpack scripts or at the least had not used them.  I will do research into these today.  Thanks

**Comment by hawkeye64** [comment]
@markbruk You can use `require.context(videoFolder, true, /^\.\/.*\.mp4$/)` to return an array of your available videos at build time.

**Reply by ontwikkelfabriek** [reply]
@hawkeye64 

> You can use require.context(videoFolder, true, /^\.\/.*\.mp4$/)

Where and how would you store this info so it's available at runtime?

**Comment by hawkeye64** [comment]
@ontwikkelfabriek 

> Where and how would you store this info so it's available at runtime?

If you look at this file I import both `menu.js`, `articles.js` and `examples.js` which do something similar, Then I dynamically create routes (for examples): https://github.com/quasarframework/quasar-ui-qcalendar/blob/next/docs/src/router/routes.js
